docs: update developer reference docs (#126)
matter-labs/zksync-docs: content/00.build/65.developer-reference/50.era-contracts/10.l1-contracts.md

---
::

This section delves into the ZKsync Era smart contracts,
describing how they facilitate interactions between Ethereum (Layer 1)
and hyperchain instances (Layer 2) within ZKsync Era ecosystem.
While we provide overview of ZKsync Era smart contracts here,

> it's important to note that this document does *NOT- cover the Hyperchain architecture -

Reviewer: ⚠️⚠️⚠️  IMPORTANT ⚠️⚠️⚠️
We need to replace all instances of the verbiage `Hyperchain` with `ZK Chain` (per legal)

Reviewer: Also unsure if the `-` are intended, perhaps update to the following?
```suggestion
it's important to note that this document does *NOT* cover the ZK Chain architecture
```

communication between multiple rollups and shared liquidity in details.
For information on the Hyperchain and its functionalities,
please refer to [Ecosystem contracts](https://github.com/code-423n4/2024-03-zksync/blob/main/docs/Smart%20contract%20Section/L1%20ecosystem%20contracts.md).

> <!-- TODO: update ecosystem contracts link -->

Reviewer: Does this still need to be done? I know v23 branch was merged in `era-contracts` already, so should be good to point to `main`

## Diamond (also mentioned as State Transition contract)

Technically, this L1 smart contract acts as a connector between Ethereum (L1) and hyperchain (L2).
It checks the validity proof and data availability,
handles L2 <-> L1 communication, finalizes L2 state transition, and more.

> There are also important contracts deployed on the L2 that can also execute logic called *system contracts*.

Reviewer: nit: `system contracts` should probably link to the System Contracts page
```suggestion
There are also important contracts deployed on the L2 that can also execute logic called [system contracts](/build/developer-reference/era-contracts/system-contracts).
```

implementation that is inspired by the [mudgen reference implementation](https://github.com/mudgen/Diamond). It has no
external functions, only the fallback that delegates a call to one of the facets (target/implementation contract). So
even an upgrade system is a separate facet that can be replaced.

One of the differences from the reference implementation is access freeze-ability. Each of the facets has an associated
parameter that indicates if it is possible to freeze access to the facet. Privileged actors can freeze the **diamond**
(not a specific facet!) and all facets with the marker `isFreezable` should be inaccessible until the governor or admin
unfreezes the diamond. Note that it is a very dangerous thing since the diamond proxy can freeze the upgrade system and then
the diamond will be frozen forever.

The diamond proxy pattern is very flexible and extendable.
For now, it allows splitting implementation contracts by their logical meaning,
removes the limit of bytecode size per contract and implements security features such as freezing.
In the future, it can also be viewed as [EIP-6900](https://eips.ethereum.org/EIPS/eip-6900)

> for [zkStack](https://blog.matter-labs.io/introducing-the-zk-stack-c24240c2532a),

Reviewer: All `zkStack` verbiage needs to be updated to `ZK Stack`
```suggestion
for [ZK Stack](https://blog.matter-labs.io/introducing-the-zk-stack-c24240c2532a),
```

where each hyperchain can implement a sub-set of allowed implementation contracts.

### GettersFacet

Separate facet, whose only function is providing `view` and `pure` methods. It also implements

> [diamond loupe](https://eips.ethereum.org/EIPS/eip-2535#diamond-loupe) which makes managing facets easier.

Reviewer: nit: The specific text `diamond-loupe` does not exist in the document. Perhaps we should use the following direct link instead? https://eips.ethereum.org/EIPS/eip-2535#inspecting-facets--functions

Reviewer 2: Good catch, yes the link should be fixed

This contract must never be frozen.

### AdminFacet

This facet responsible for the configuration setup and upgradeability, handling tasks such as:

- Privileged Address Management: Updating key roles, including the governor and validators.
- System Parameter Configuration: Adjusting critical system settings,
  such as the L2 bootloader bytecode hash, verifier address, verifier parameters, fee configurations.
- Freeze-ability: Executing the freezing/unfreezing of facets within the diamond proxy
  to safeguard the ecosystem during upgrades or in response to detected vulnerabilities.

Control over the AdminFacet is divided between two main entities:

- STM (State Transition Manager) - Separate smart contract that can perform critical changes to the system as protocol upgrades.
  For more detailed information on its function and design, refer to the

>   [Hyperchain section](https://github.com/code-423n4/2024-03-zksync/blob/main/docs/Smart%20contract%20Section/L1%20ecosystem%20contracts.md#st--stm).

Reviewer: Should this link to the internal docs page we are migrating?

Reviewer: This happens a lot in the documents added, we should address accordingly

  Although currently only one version of the STM exists, the architecture allows for future versions to be introduced via subsequent upgrades.
  Control of the STM is shared between the `Governance.sol` contract and the Admin entity (see details below).
  In its turn, `Governance.sol` controlled by two multisigs:
  Admin multisig (see below) and Security council multisig (well-respected contributors in the crypto space).
  Collaboratively, these entities hold the power to implement instant upgrades,
  whereas Matter Labs alone is limited to scheduling upgrades with a delay.
- Admin - Multisig smart contract managed by Matter Labs that can perform non-critical
  changes to the system such as granting validator permissions.
  Note, that the Admin is the same multisig as the owner of the governance.

### MailboxFacet

> > **NOTE:** This contract significantly changed from the time of the previous code4rena competition.

Reviewer: Do we still want to include notes like this that are very specific to `code4rena`?
